chain-abci/src/enclave_bridge/mod.rs

```diff
@@ -1,40 +1,15 @@
-use enclave_protocol::{EnclaveRequest, EnclaveResponse, FLAGS};
+use enclave_protocol::{IntraEnclaveRequest, IntraEnclaveResponse};
 use parity_scale_codec::{Decode, Encode};
-use std::sync::{Arc, Mutex};
-use zmq::Socket;
 
 /// TODO: feature-guard when workspaces can be built with --features flag: https://github.com/rust-lang/cargo/issues/5015
 pub mod mock;
 
-/// Abstracts over communication with an external process that does enclave calls
-pub trait EnclaveProxy: Sync + Send + Sized {
-    fn process_request(&mut self, request: EnclaveRequest) -> EnclaveResponse;
-}
-
-/// Provides communication with the enclave wrapper app over ZMQ
-/// NOTE / WARNING: this connection is trusted / non-attested
-/// (it's assumed Tendermint node, Chain ABCI app and enclave process would run on the same machine)
-pub struct ZmqEnclaveClient {
-    socket: Arc<Mutex<Socket>>,
-}
+#[cfg(target_os = "linux")]
+pub mod real;
 
-impl ZmqEnclaveClient {
-    pub fn new(socket: Socket) -> Self {
-        ZmqEnclaveClient {
-            socket: Arc::new(Mutex::new(socket)),
-        }
-    }
-}
-
-impl EnclaveProxy for ZmqEnclaveClient {
-    fn process_request(&mut self, request: EnclaveRequest) -> EnclaveResponse {
-        let asocket = Arc::clone(&self.socket);
-        let socket = asocket.lock().unwrap();
-        let req = request.encode();
-        socket.send(req, FLAGS).expect("request sending failed");
-        let msg = socket
-            .recv_bytes(FLAGS)
-            .expect("failed to receive a response");
-        EnclaveResponse::decode(&mut msg.as_slice()).expect("failed to parse a response")
-    }
+/// Abstracts over communication with an external part that does enclave calls
+pub trait EnclaveProxy: Sync + Send + Sized {
```

Comment:
Awesome! This is much cleaner and easy to understand. Can we also split variants of `IntraEnclaveRequest` into trait functions (i.e., have three functions, `end_block`, `encrypt` and `validate_tx`, instead of single `process_request` function)?

Reply:
I'd delay that in the PR with fixes to the enclave protocol.
One other thing is that ValidateTx is kind of "one size fits all" variant and it'll be cleaner to break it down, so only valid requests can be constructed + only meaningful return types are returned

Reply:
Ok.